src/amd64/amd64.c.inc

```diff
@@ -0,0 +1,201 @@
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Parts of this file from biench.cr.yp.to/supercop.html (2017-02-25):
+//     Daniel J. Bernstein
+//     Niels Duif
+//     Tanja Lange
+//     lead: Peter Schwabe
+//     Bo-Yin Yang
+
+#include <string.h>
+#include "ge25519.h"
+
+extern void choose_tp(ge25519_pniels *t, unsigned long long pos, signed long long b, const ge25519_pniels *base_multiples);
+
+/* return 0 on success, -1 otherwise. Taken from
+ed25519/amd64-51-30k/ge25519_unpackneg.c - the negation is removed. */
+static int unpack_vartime(ge25519_p3 *r, const unsigned char p[32])
+{
+  fe25519 t, chk, num, den, den2, den4, den6;
+  unsigned char par = p[31] >> 7;
+
+  fe25519_setint(&r->z,1);
+  fe25519_unpack(&r->y, p); 
+  fe25519_square(&num, &r->y); /* x = y^2 */
+  fe25519_mul(&den, &num, &ecd); /* den = dy^2 */
+  fe25519_sub(&num, &num, &r->z); /* x = y^2-1 */
+  fe25519_add(&den, &r->z, &den); /* den = dy^2+1 */
+
+  /* Computation of sqrt(num/den)
+     1.: computation of num^((p-5)/8)*den^((7p-35)/8) = (num*den^7)^((p-5)/8)
+  */
+  fe25519_square(&den2, &den);
+  fe25519_square(&den4, &den2);
+  fe25519_mul(&den6, &den4, &den2);
+  fe25519_mul(&t, &den6, &num);
+  fe25519_mul(&t, &t, &den);
+
+  fe25519_pow2523(&t, &t);
+  /* 2. computation of r->x = t * num * den^3
+  */
+  fe25519_mul(&t, &t, &num);
+  fe25519_mul(&t, &t, &den);
+  fe25519_mul(&t, &t, &den);
+  fe25519_mul(&r->x, &t, &den);
+
+  /* 3. Check whether sqrt computation gave correct result, multiply by sqrt(-1) if not:
+  */
+  fe25519_square(&chk, &r->x);
+  fe25519_mul(&chk, &chk, &den);
+  if (!fe25519_iseq_vartime(&chk, &num))
+    fe25519_mul(&r->x, &r->x, &sqrtm1);
+
+  /* 4. Now we have one of the two square roots, except if input was not a square
+  */
+  fe25519_square(&chk, &r->x);
+  fe25519_mul(&chk, &chk, &den);
+  if (!fe25519_iseq_vartime(&chk, &num))
+    return -1;
+
+  /* 5. Choose the desired square root according to parity:
+  */
+  if(fe25519_getparity(&r->x) == (1-par)) // only change from original function was `!=` -> `==`
+    fe25519_neg(&r->x, &r->x);
```

Comment:
The difference from the original here is:

```
   /* 5. Choose the desired square root according to parity:
   */
-  if(fe25519_getparity(&r->x) != (1-par))
+  if(fe25519_getparity(&r->x) == (1-par)) // only change from original function was `!=` -> `==`
     fe25519_neg(&r->x, &r->x);
 
   fe25519_mul(&r->t, &r->x, &r->y);
```

No clue why, so I'm putting this here and will ask luigi.

Reply:
@luigi1111 

The Ed25519 implementations unpack and negate the point at the same time. This is because public-keys are only used to [verify a signature](https://github.com/monero-project/supercop/blob/0d6dff95350a20f6fe2bafe1ff6730458dd5ddef/crypto_sign/ed25519/amd64-64-24k/open.c), which is an implementation of the [EdDSA signature equation](https://en.wikipedia.org/wiki/EdDSA). The function is subtracting the value on the far right from the left side of the equation to compare equality to the point `R`.

This function isn't strictly needed, but the x-coordinate would need to be immediately negated before calculating the ECDH for the transaction, etc. So this is for optimization purposes.

Reply:
@vtnerd ah, thanks.